### ISSUE TITLE
[TW-440] pending.data corrupted

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#475]
_Bryce Harrington on 2012-09-25T22:36:02Z says:_ 

 Not entirely sure how this happened, but I ran `task project:work.xorg.multihead` then `task project:work.xorg.multihead all`.  I think I may have run the latter command twice and might have ^C'd one or the other.

Anyway, it produced an error something to the extent that pending.data was in old format 1, no longer supported.  But looking at the pending.data file, actually the file was corrupted.  Here's an example snippet:

[description:"Write a short newbie-oriented tutorial on arsenal website" entry:"1307612833" priority:"L" project:"work.arsen
al" status:"pe[description:"Shop for a new HDTV -  http:\/\/www.pc[description:"Make script to find reports of non-distro ve
rsions of nvidia and close as wontfix" entry:"1308378476" priority:"L" project:"work.arsenal.future" status:"pending" uuid:"
d657cc2e-2fac-1c[description:"In xdiagn[description:"Make script to identify nvidia bug reports that have been reopened" ent
ry:"1308378490" priority:"L" project:"work.arsenal.future" status:"pending" uuid:"ab1c67[description:"For xdiagnose consi[de
scription:"Add support for removing monitors.xml" entry:"1308378559" priority:"L" project:"work.xdiagnose.proto" status:"pen
ding" uuid:"53[description:"In xdiagnose add functi[description:"Add support for installing\/launching driconf" entry:"13083
78633" priority:"L" project:"work.xdiagnose.future" status:"pending" uuid:"4698b18f-408f-4e27-cc24-bb32ab9f763b"]
d5-9187-b86e-0de3-ff083f97865e"]
[description:"Add workloads tab with checkboxes for various stressors to run" entry:"1308378665" priority:"M" project:"work.
xdiagnose.proto" status:"pending" uuid:"0774ef0f-2d1f-1a56-65c3-097fef04f4df"]
e3f704407992"]
[annotation_1313707681:"Notes" description:"functionality to turn on mmiotrace" entry:"1308378685" priority:"L" project:"wor
k.xdiagnose.future" status:"pending" uuid:"c5a56b1c-9ef2-a9ba-e26b-dcee1401210f"][[description:"Create ncurses alternative f
rontend" entry:"1308378825" priority:"L" project:"work.xdiagnose.future" status:"pending" uuid:"e8db779e-f5ba-eaf6-aa31-b4af
f237bef9"]
07d-509b-d7958f38085b"]

This is with task 2.0.0 (Ubuntu 12.04's package).  I did not try against 2.1.2, however looking at the git log between 2.0.0 and 2.1.2 I didn't spot an obvious "Fixes data corruption problem" entry so guessing it might affect 2.1.2 as well.

Additional background, I store my *.data files on an NFSv3 share from a different machine, and access them locally, with a local ~/.taskrc file.  99% of the time I use the taskhelm interface to interact with tasks, but this particular time I was not using taskhelm, just the task cli client, and no other processes would have been reading or writing these files afaik.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2012-09-25T22:36:02Z
Modified: 2016-01-26T18:57:03Z
```

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2012-09-25T23:03:01Z says:_ 

 Going through the bad pending.data breakage, I'm wondering if it was trying to update recurring tasks and just failed to lock properly or something?

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2012-09-25T23:52:14Z says:_ 

 How I reconstructed the data (I had some backups but they were old; I think this gets most of my current stuff back):

1.  Back everything up (again)
2.  In pending.data replace all "[" with "\n[" so each task/annotation fragment is on its own line
3.  grep "^\[.*\]$" pending.data >pending.data.good
4.  grep -v "^\[.*\]$" pending.data >pending.data.remainders
5.  mv pending.data.good pending.data
6.  For each non-blank line in pending.data, grep from undo.data; replace line in pending.data with the last appropriate match from undo.data
7.  Merge results from step 6 into pending.data from step 5

If there's a more robust way of rebuilding pending.data, let me know.  I suspect the process above is a bit lossy.

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2012-09-26T01:27:18Z says:_ 

 I think I might know what happened.  I recall last night I was working on a different computer (which uses nfs over wireless) and marked six recurring tasks as done in taskhelm.  However, it failed to remove those tasks from the list.  I ignored that and went to another computer, and closed the tasks there.  I wonder if perhaps the first one delayed its writes, and the second one stepped on its toes.  If that's the case, there may not be anything that can be done for task here.  Not sure how to prove that's what happened, let alone reproduce it, but if you think that sounds like what happened you can close this bug report out.

Attached is my recovery script from step 6.  Even if this is some weird misbehavior, if anyone else gets file corruption like me, they might find this useful.  Note it has some paths hard coded.  After running this, I concatenated the two files and then to remove duplicate lines did `awk ' !x[$0]++' pending.data.combined > pending.data`

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2012-09-26T01:27:18Z says:_ 

 Attachment task-rebuild has been added with description: Task recovery script

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-09-27T02:43:40Z says:_ 

 Bryce, does the use of NFS mean that file locking was turned off?  Curious.

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2012-09-27T08:38:00Z says:_ 

 $ rpcinfo -p chideok | grep lock
    100021    1   udp  53750  nlockmgr
    100021    3   udp  53750  nlockmgr
    100021    4   udp  53750  nlockmgr
    100021    1   tcp  50647  nlockmgr
    100021    3   tcp  50647  nlockmgr
    100021    4   tcp  50647  nlockmgr

So no, locking is enabled.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2013-02-11T04:29:54Z says:_ 

 News: Bug #1137 caused me to look closely at locking, and it was mis-implemented, such that all file locks were non-blocking.  This means the flock() calls return immediately, but the calling code assumed blocking, and proceeded as though it had locked the file.  #1137 is fixed - perhaps this is addressed too.

Bryce, do you see a reason to keep this bug open now?

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2013-02-11T19:35:27Z says:_ 

 No, I think you can close it.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2013-02-12T01:11:37Z says:_ 

 Likely fixed by #1137.

**Comment by pbeckingham** [comment]
[TW-440_task-rebuild.txt](https://github.com/GothenburgBitFactory/taskwarrior/files/1756120/TW-440_task-rebuild.txt)
